src/digraph.rs

```diff
@@ -2889,6 +2889,48 @@ impl PyDiGraph {
         }
     }
 
+    /// Filters a graph's nodes by some criteria conditioned on a node's data payload and returns those nodes' indices
+    ///
+    /// :param filter_function: Function with which to filter nodes
+    /// :returns: The node indices that match the filter
+    /// :rtype: NodeIndices
+    #[pyo3(text_signature = "(self, filter_function)")]
+    pub fn filter_nodes(&self, py: Python, filter_function: PyObject) -> PyResult<NodeIndices> {
+        let filter = |nindex: NodeIndex| -> PyResult<bool> {
+            let res = filter_function.call1(py, (&self.graph[nindex],))?;
+            res.extract(py)
+        };
+
+        let mut n = Vec::new();
```

Comment:
A potential performance improvement we can make here is to do:

```suggestion
        let mut n = Vec::with_capacity(self.graph.node_count());
```

This basically allocates up front a `Vec` with enough space for the number of nodes in the graph. THis avoids having to do multiple memory allocations as the `Vec` grows (which are slow). The tradeoff here being that we are likely over allocating memory in the return object (unless `filter_function` returns `True` for every node. I'm thinking this tradeoff is ok in this case though because each element of the output array is just a 64bit integer so except for pathological edge cases where we have millions of nodes and the filter only has a few matches nobody will really notice if we've allocated and extra space for a few integers.

But, I'm curious what your thoughts are on this tradeoff and what you think we should do here?

Reply:
ahh, this makes sense. I think over allocating is worth it in the long run, especially if reallocating is expensive. I'll make that change!